src/mem/ruby/protocol/GPU_VIPER-TCP.sm

```diff
@@ -687,14 +737,38 @@ machine(MachineType:TCP, "GPU TCP (L1 Data Cache)")
     p_popMandatoryQueue;
   }
 
-  transition(I, TCC_Ack, V) {TagArrayRead, TagArrayWrite, DataArrayRead, DataArrayWrite} {
+  // if we got a response for a load where the line is in I, then
+  // another request must have come in that replaced the line in question in
+  // the cache.  Thus, complete this request without allocating the line, but
+  // still deallocate TBE and wakeup any dependent addresses.
+  // (Note: this assumes TCC_AckWB is what stores use)
+  transition(I, TCC_Ack) {TagArrayRead, TagArrayWrite} {
+    pr_popResponseQueue;
+    wada_wakeUpAllDependentsAddr;
+    // NOTE: Because we invalidated the cache line, the assert in l_loadDoneMiss
+    // will fail -- unlike atomics that automatically go to I when the line returns
+    // loads do not automatically go to I.  Resolve this by passing data from
+    // message.
+    ldmi_loadDoneMissInv;
+    d_deallocateTBE;
+  }
```

Comment:
Perhaps the seg fault is if you have two TCC_Ack responses or something?  I don't really have any other ideas.

I don't see a good way around this, but a line in a transient state shouldn't be selected as a victim IMO.  It doesn't appear that cacheProbe considers the "AccessPermission" of a line though. Maybe it just works out that the replacement policy only picks addresses in base cases because they are the oldest.

Really it seems like if there are *no* lines in a base state that can be replaced, we should stall the incoming requests until there is one.  TCP looks like it is not allocating until a response is received, so I guess you would have to count the number of in-flight requests to a cache set and stall further requests if that is >= the associativity.  That's a big change though, so probably finding the seg fault is going to be easier.